### PR TITLE
gh-114053: Fix bad interaction of PEP-695, PEP-563 and ``get_type_hints``

### PR DESCRIPTION
**Pull request by AlexWaygood**
* Issue: gh-114053

### CODE REVIEW

File: Lib/typing.py [hunk 1]
```diff
@@ -989,13 +996,21 @@ def _evaluate(self, globalns, localns, recursive_guard):
                     sys.modules.get(self.__forward_module__, None), '__dict__', globalns
                 )
             type_ = _type_check(
-                eval(self.__forward_code__, globalns, localns),
+                eval(
+                    self.__forward_code__,
+                    globalns,
+                    {param.__name__: param for param in type_params} | localns
```
Comment (AlexWaygood) [inline Lib/typing.py]: this means that any annotations in the "local namespace" of the class override any symbols that are present in the scope due to being listed as a type parameter. That matches the behaviour of `__annotations__` at runtime in a module that doesn't use `__future__` annotations:

```pycon
>>> class B[T, *Ts, **P]:
...     T = int
...     Ts = str
...     P = bytes
...     x: T
...     y: Ts
...     z: P
... 
>>> B.__annotations__
{'x': <class 'int'>, 'y': <class 'str'>, 'z': <class 'bytes'>}
```

I'm doing it at the last possible point here as it feels less risky than doing it in `_eval_type`.

File: Doc/library/typing.rst [hunk 2]
```diff
@@ -3024,7 +3024,9 @@ Introspection helpers
 
    This is often the same as ``obj.__annotations__``. In addition,
    forward references encoded as string literals are handled by evaluating
-   them in ``globals`` and ``locals`` namespaces. For a class ``C``, return
+   them in ``globals``, ``locals`` and (where applicable)
+   :ref:`type parameter <annotation-scopes>` namespaces.
+   For a class ``C``, return
```
Comment (AlexWaygood) [inline Doc/library/typing.rst]: we don't usually add `.. versionchanged` notes for bugfixes (and I think this is a bugfix), but I could add one if we think this is a significant enough change to the function's semantics

File: Doc/library/typing.rst [hunk 3]
```diff
@@ -3024,7 +3024,9 @@ Introspection helpers
 
    This is often the same as ``obj.__annotations__``. In addition,
    forward references encoded as string literals are handled by evaluating
-   them in ``globals`` and ``locals`` namespaces. For a class ``C``, return
+   them in ``globals``, ``locals`` and (where applicable)
+   :ref:`type parameter <annotation-scopes>` namespaces.
```
Comment (JelleZijlstra) [inline Doc/library/typing.rst]: This should link to https://docs.python.org/3.12/reference/compound_stmts.html#type-params instead. Annotation scopes are a more general concept.
```suggestion
   :ref:`type parameter <type-params>` namespaces.
```

**Review by JelleZijlstra — APPROVED** [review]
Thanks, this is a much simpler fix than I was afraid we'd need.

**Comment by dimaqq** [comment]
cc for 3.12 backport:

https://discuss.python.org/t/3-12-3-3-12-4-regression-deep-rabbit-hole-warning/59584/12

**Pull request merged**